gotham/src/handler/assets/mod.rs

```diff
@@ -194,59 +198,62 @@ impl Handler for DirHandler {
 }
 
 impl Handler for FileHandler {
-    fn handle(self, state: State) -> Box<HandlerFuture> {
+    fn handle(self, state: State) -> Pin<Box<HandlerFuture>> {
         create_file_response(self.options, state)
     }
 }
 
 // Creates the `HandlerFuture` response based on the given `FileOptions`.
-fn create_file_response(options: FileOptions, state: State) -> Box<HandlerFuture> {
+fn create_file_response(options: FileOptions, state: State) -> Pin<Box<HandlerFuture>> {
     let mime_type = mime_for_path(&options.path);
     let headers = HeaderMap::borrow_from(&state).clone();
 
     let (path, encoding) = check_compressed_options(&options, &headers);
 
-    let response_future =
-        File::open(path)
-            .and_then(File::metadata)
-            .and_then(move |(file, meta)| {
-                if not_modified(&meta, &headers) {
-                    return Ok(http::Response::builder()
-                        .status(StatusCode::NOT_MODIFIED)
-                        .body(Body::empty())
-                        .unwrap());
-                }
-                let len = meta.len();
-                let buf_size = optimal_buf_size(&meta);
-
-                let stream = file_stream(file, buf_size, len);
-                let body = Body::wrap_stream(stream);
-                let mut response = http::Response::builder();
-                response.status(StatusCode::OK);
-                response.header(CONTENT_LENGTH, len);
-                response.header(CONTENT_TYPE, mime_type.as_ref());
-                response.header(CACHE_CONTROL, options.cache_control);
-
-                if let Some(etag) = entity_tag(&meta) {
-                    response.header(ETAG, etag);
-                }
-                if let Some(content_encoding) = encoding {
-                    response.header(CONTENT_ENCODING, content_encoding);
-                }
-
-                Ok(response.body(body).unwrap())
-            });
-    Box::new(response_future.then(|result| match result {
-        Ok(response) => Ok((state, response)),
-        Err(err) => {
-            let status = match err.kind() {
-                io::ErrorKind::NotFound => StatusCode::NOT_FOUND,
-                io::ErrorKind::PermissionDenied => StatusCode::FORBIDDEN,
-                _ => StatusCode::INTERNAL_SERVER_ERROR,
-            };
-            Err((state, err.into_handler_error().with_status(status)))
+    let response_future = File::open(path).and_then(|file| {
+        async move {
```

Comment:
For my sake, because I'm not completely up to date with the syntax... can this become `|file| async move {`?

Reply:
Unfortunately no.  Async closures [are not stable yet][0].  Oddly, you can pass a closure and an async block in lieu.

[0]: https://rust-lang.github.io/rfcs/2394-async_await.html#async--closures